### DISCUSSION TITLE
A question about a symbol in the C# grammar

### DISCUSSION THREAD

**Discussion by vsajip** [javacc21/javacc21#185] [Q&A] [answered]
Is there a particular reason this symbol has to start with a double-underscore? 
https://github.com/javacc21/javacc21/blob/ba4f192a8f7b07af6981530a03074e1004baec71/examples/csharp/CSharpLexer.javacc#L186
I ask because it causes problems in Python - identifiers with double-underscores are either magic methods or [class-private constants](https://docs.python.org/3/faq/programming.html#i-try-to-use-spam-and-i-get-an-error-about-someclassname-spam). I'd like to change this to just plain `ARGLIST`.

**Comment by revusky** [comment, marked as answer]
Well, the reason is that it should be:

      <__ARGLIST : "__arglist">

The "soft keyword" is with the two underscores in front. The reason I never noticed that is because, for some reason, I defined the contextual keyword but then never used it! If you look at the [ArgListParameter production](https://github.com/javacc21/javacc21/blob/master/examples/csharp/CSharp.javacc#L762) for some reason it just matches it as an identifier. Maybe there was some reason I decided to do it that way, but I honestly can't remember. Most likely I had originally written it that way and then decided to define the contextual keyword, but then never got round to patching the grammar to actually use that, and so that is why my writing the token definition wrong was never caught!

And, of course, there is no particular reason for the name of the token to start with two underscores, so I ended up changing it to:

       <ARGLIST : "__arglist">

By the way, I intend to streamline the syntax of all this at some point soon. Something like this, using single quotes in these spots. So, now, you might write something like:

      SCAN {checkNextTokenImage("__arglist")} ==> ACTIVATE_TOKENS ARGLIST ("__arglist")

which is kind of ridiculous, really, I know. Though it does make the machinery clear, I guess. It scans ahead, and since the ARGLIST token type is not active, it matches the "__arglist" string as an IDENTIFIER, but then since it matches the string in question, it throws away that IDENTIFIER token (which is implicit in the ACTIVATE_TOKENS thing) and then matches the same in put as the ARGLIST type.

BUT.... I think we can see our way clear to replacing the above line with:

     '__arglist'

Just using single quote. I suppose that if there is a contextual token defined with that string, then it does the above. And if you do just want to match this as an IDENTIFIER, say, then the above just would be equivalent to:

    <IDENTIFIER> 
    ASSERT {getToken(0).getImage().equals("__arglist")}#

I've been meaning to do that for ages now. I mean, the current thing works okay, but is really too verbose.

**Comment by vsajip** [comment]
>  <ARGLIST : "__arglist">

I tried that and it worked for me. I still have a way to go to provide a working C# parser  generated in Python (and then will move to the parser generated in C#).

**Comment by revusky** [comment]
> I still have a way to go to provide a working C# parser generated in Python (and then will move to the parser generated in C#).

I'm thinking that when we do have a C# parser generated in C# itself, we should really open a line of communication with the C# developers. I would also like to have the code injections and actions working better for both C# and Python. But maybe it would make sense to start talking with the relevant people sooner rather than later. 

It will take some thought about how to approach the situation, I guess. But we will meet in person in a few weeks and maybe hash it out.

**Reply by vsajip** [reply]
> I would also like to have the code injections and actions working better for both C# and Python

The smoothest development process will occur (with the transpiling approach) when code in injections, token hooks etc. are written in a way to facilitate portability across platforms. For example, I've had to rework code that uses the JDK streams API as other languages and frameworks do these things differently.
